static/js/narrow.js

```diff
@@ -295,10 +292,34 @@ export function activate(raw_operators, opts) {
             // the stream/topic pair that was requested to some other
             // location, then we should retarget this narrow operation
             // to where the message is located now.
-            const adjusted_operators = adjusted_operators_if_moved(raw_operators, target_message);
-            if (operators_changed) {
-                activate(adjusted_operators, opts);
-                return;
+            const narrow_topic = filter.operands("topic")[0];
+            const narrow_stream_name = filter.operands("stream")[0];
+            const narrow_stream_id = stream_data.get_sub(narrow_stream_name).stream_id;
+            const narrow_dict = {stream_id: narrow_stream_id, topic: narrow_topic};
+
+            const narrow_matches_target_message = util.same_stream_and_topic(
+                target_message,
+                narrow_dict,
+            );
+            const narrow_exists_in_edit_history =
+                message_edit.stream_and_topic_exist_in_edit_history(
+                    target_message,
+                    narrow_stream_id,
+                    narrow_topic,
+                );
+
+            if (
+                !narrow_matches_target_message &&
+                (narrow_exists_in_edit_history || !page_params.realm_allow_edit_history)
+            ) {
+                const adjusted_operators = adjusted_operators_if_moved(
+                    raw_operators,
+                    target_message,
+                );
+                if (operators_changed) {
```

Comment:
I think we're missing something that sets this in what `adjusted_operators_if_moved` returns. I think for case-sensitivity and ordering reasons, we do want that to be something that is explicitly returned rather than trying to compare `adjusted_operators` to `raw_operators`, so maybe `adjusted_operators_if_moved` should return an object to support that.

Reply:
I'm having trouble following your comment/thought here. I don't see where we're comparing `adjusted_operators` to `raw_operators`.

`operators_changed` is set as `false` before `adjusted_operators_if_moved` is defined. But then, if/when that function is called, and if it does redirect the narrow and created a new set of operators, then it's changed to `true`.

Are you saying that instead of that behavior, `adjusted_operators_if_moved` should return an object with a boolean and an array. If the boolean is false, then the array will be a copy of `raw_operators` and we don't do the recursive call. If the boolean is true, then we have an array with the updated operators (`adjusted_operators`) to use in the recursive call to `activate`? And we can then delete the preset `operators_changed` and `adjusted_operators`?

Reply:
Oh, sorry, we are not comparing those at present -- the point was that if we were trying to remove the need for a separate `operators_changed` variable, checking if the adjusted operators matched the original operators would be a way to try to do that ... but it doesn't actually work.

Now that I think about it, though, we could just make `adjusted_operators_if_moved` return `null` if it didn't identify any changes, and then not need the object model.